geozero/src/gpx/gpx_reader.rs

```diff
@@ -0,0 +1,227 @@
+use crate::error::GeozeroError;
+use std::io;
+
+/// GPX reader
+pub struct GpxReader<'a, R: io::Read>(pub &'a mut R);
+
+impl<'a, R: io::Read> crate::GeozeroDatasource for GpxReader<'a, R> {
+    fn process<P: crate::FeatureProcessor>(
+        &mut self,
+        processor: &mut P,
+    ) -> crate::error::Result<()> {
+        read_gpx(&mut self.0, processor)
+    }
+}
+
+pub fn read_gpx<R: io::Read, P: crate::GeomProcessor>(
+    reader: &mut R,
+    processor: &mut P,
+) -> crate::error::Result<()> {
+    let gpx_reader = match gpx::read(reader) {
+        Ok(r) => r,
+        Err(e) => return Err(GeozeroError::Geometry(e.to_string())),
+    };
+
+    process_top_level_waypoints(&gpx_reader, processor)?;
+    process_top_level_tracks(&gpx_reader, processor)?;
+    process_top_level_routes(&gpx_reader, processor)?;
+
+    Ok(())
+}
+
+fn process_top_level_waypoints<P: crate::GeomProcessor>(
+    gpx_reader: &gpx::Gpx,
+    processor: &mut P,
+) -> crate::error::Result<()> {
+    if gpx_reader.waypoints.is_empty() {
+        return Ok(());
+    }
+    processor.multipoint_begin(gpx_reader.waypoints.len(), 0)?;
+    process_waypoints_iter(gpx_reader.waypoints.iter(), processor)?;
+    processor.multipoint_end(0)?;
+    Ok(())
+}
+
+fn process_top_level_tracks<P: crate::GeomProcessor>(
+    gpx_reader: &gpx::Gpx,
+    processor: &mut P,
+) -> crate::error::Result<()> {
+    if gpx_reader.tracks.is_empty() {
+        return Ok(());
+    }
+    for (index, track) in gpx_reader.tracks.iter().enumerate() {
+        process_track_segments(track, processor, index)?;
+    }
+    Ok(())
+}
+
+fn process_track_segments<P: crate::GeomProcessor>(
+    track: &gpx::Track,
+    processor: &mut P,
+    index: usize,
+) -> crate::error::Result<()> {
+    if track.segments.is_empty() {
+        return Ok(());
+    }
+    processor.multilinestring_begin(track.segments.len(), index)?;
+    for (index, segment) in track.segments.iter().enumerate() {
+        process_track_segment(segment, processor, index)?;
+    }
+    processor.multilinestring_end(index)?;
+    Ok(())
+}
+
+fn process_track_segment<P: crate::GeomProcessor>(
+    segment: &gpx::TrackSegment,
+    processor: &mut P,
+    index: usize,
+) -> crate::error::Result<()> {
+    if segment.points.is_empty() {
+        return Ok(());
+    }
+    processor.linestring_begin(false, segment.points.len(), index)?;
+    process_waypoints_iter(segment.points.iter(), processor)?;
+    processor.linestring_end(false, index)?;
+    Ok(())
+}
+
+fn process_top_level_routes<P: crate::GeomProcessor>(
+    gpx_reader: &gpx::Gpx,
+    processor: &mut P,
+) -> crate::error::Result<()> {
+    if gpx_reader.routes.is_empty() {
+        return Ok(());
+    }
+    processor.multilinestring_begin(gpx_reader.routes.len(), 0)?;
```

Comment:
Is it OK to always pass 0 here? IIRC CSV and maybe some other formats use the index to determine when to have a separating comma.

Reply:
Looking at `test_extensive` the output is a `GeometryCollection` with one `MultiPoint` (index 0) and two `MultiLineString` geometries (index 1+2).

Reply:
The tests are using the `geo-types` writer, which doesn't use the index parameter, so you could give it any value you wanted. The parameter exists because other writers/processors do use this parameter.

Reply:
I guess let me see if I can break it. =)

Reply:
As a comparison: `ogrinfo` shows 5 layers with the following geometry types:
- 1: waypoints (Point)
- 2: routes (Line String)
- 3: tracks (Multi Line String)
- 4: route_points (Point)
- 5: track_points (Point)

As we don't have multiple layers, we can have a collection containing the 3 first layers. I would not issue a MultiPoint for waypoints, since these points don't belong to a single feature.

Reply:
I introduced a `TestWriter` in my latest commit that better visualizes all the processor commands that are run: https://github.com/georust/geozero/pull/48/commits/c50a32f2629ba974bf1a293e37f85d43a34721b5

Regarding the GPX changes you both are talking about, I've lost the thread a little. What should I be changing? It sounds like waypoints should not be a `MultiPoint`.

Reply:
> I guess let me see if I can break it.

See #51 for some executable examples of issues I see with the current implementation.

Feel free to build off of it or just use it as a reference, and let me know if you think any of those tests are invalid.